### ISSUE TITLE
Is it possible to do batch parallelization using the  predict() model class method?

### ISSUE THREAD

**Issue by Vaevin** [ultralytics/ultralytics#1310]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I understand that we can call the model.predict() and pass in an image or even a list of images or folder path as source, for example:

```
model = YOLO("model.pt")
im1 = Image.open("bus1.jpg")
im2 = Image.open("bus2.jpg")
im_list = [im1, im2]
results = model.predict(source=im_list )
```

However, I am wondering on how to utilize batch parallelization so that I can process both images in parallel on the GPU? I've found that the trainer has a parameter for batch size (shown in figure below):

![Screenshot 2023-03-07 145024](https://user-images.githubusercontent.com/60018242/223536991-147bba32-e275-4f6e-ab9f-d1cf68cb11bf.png)

However, I do not see this feature in the predict method. In conclusion, I am wondering if there is a way to utilize this batch parallelization feature with the predictor method.

**Comment by glenn-jocher** [comment]
@Vaevin no not currently. Unless you are running multiple streaming sources then predict will run at batch-size 1. 

We have expanding batched prediction for additional source types in our product roadmap though.

@AyushExel @Laughing-q good feature request here, we've heard this also several times in YOLOv5.

**Comment by glenn-jocher** [comment]
@Vaevin forgot to mention, if you're open to contributing, PRs are welcome :)

**Comment by Vaevin** [comment]
@glenn-jocher Thank you for the quick reply.

I found in the YOLO v8 docs (very helpful docs by the way) that it does say a source could be a torch tensor with a batch_size as first index:

![image](https://user-images.githubusercontent.com/60018242/223735438-38f3fe4b-9644-4099-810c-b253f46dfad8.png)

However, when I tried creating a torch tensor of batch=16, using the pytorch dataloader and dataset classes, and use that as source I got the error "Exception: Unsupported type encountered!". Am I correct in assuming that this is the future feature implementation you are referencing? The reason I ask is because I am wondering if the source input COULD work on a tensor, but I just need to use the custom YOLO v8 dataloader creations for it or something.

And yes, I will definitely take a swing at implementing it :) However my lack of familiarity with the YOLO code base will take time.

**Comment by Vaevin** [comment]
@Laughing-q @glenn-jocher Thank you for replying. So, I know that in YOLOv5, when I passed in an image list just as you had provided, it does perform batch inference; I am able to confirm inference speed times increases dramatically. However, YOLOv8 does not appear to support this. Here is my code to recreate the YOLOv8 case (you'll need your own image folderpath and your own model):

# Performs batch inference by passing in a list of images as source.

```
import sys
from pathlib import Path
file = Path(__file__).resolve()
root = file.parents[2]  # YOLOv5 root directory
if str(root) not in sys.path:
    sys.path.append(str(root))  # add ROOT to PATH
import cv2
import numpy as np
import os
from ultralytics import YOLO

# This init_YOLO sets up everything we need. To change any parameters, weights, filepaths, etc, you MUST edit 
# the  config_YOLO.yaml file.
model = YOLO(model=weights)

weights = # Insert weights path here
Images_folderpath = # Insert folder path path here

images = []
for img in os.listdir(Images_folderpath):
    images.append(cv2.imread(os.path.join(Images_folderpath, img)))

#%%
# Create image list
list_sz = 24
imgs = []
for i in range(list_sz):
    imgs.append(images[i])


#%% Pass in each image one at a time.
inference_times = np.empty(len(imgs))
dt_total_list = []
for img in imgs:
    predictions = model.predict(source=img, save=False, save_conf=True, max_det=1, show=False, stream=False, device=0)
    
    for prediction in predictions:
        dt_dict = prediction.speed
        dt = []
        for key, value in dt_dict.items():
            dt.append(value)
        dt_total = sum(dt)
        dt_total_list.append(dt_total)
        
# Once all 20 have been looped, add that list to our inference_times
inference_times = np.array(dt_total_list)

print(f"Average Inference Speed: {np.mean(inference_times):.2f}")

#%% Pass in the entire imgs list
inference_times = np.empty(len(imgs))
dt_total_list = []

predictions = model.predict(source=imgs, save=False, save_conf=True, max_det=1, show=False, stream=False, device=0)

for prediction in predictions:
    dt_dict = prediction.speed
    dt = []
    for key, value in dt_dict.items():
        dt.append(value)
    dt_total = sum(dt)
    dt_total_list.append(dt_total)
        
inference_times = np.array(dt_total_list)

print(f"Average Inference Speed: {np.mean(inference_times):.2f}")
```

After running the above, I see that it still only passes 1 image at a time, even when the source is an image list. The inference speeds also confirm that they are the same between passing an image list or one image at a time.

Regarding the tensor, the tensor input also does not work as a source input. 

I like the functionality of YOLOv8 and would love for it to support batch inference; I think it would be a widely regarded feature amongst the community.

**Comment by glenn-jocher** [comment]
@Vaevin yes you're right, the YOLOv5 AutoShape batches run as batched inference. We need to make some improvements here to apply the same. I'll add a TODO here to keep track of this task.

**Comment by glenn-jocher** [comment]
@Laughing-q hey good points! 

@Vaevin I stand corrected, batch inference is already in place for YOLOv8 when passed a list of images per https://github.com/ultralytics/ultralytics/issues/1310#issuecomment-1472201110. You may want to update to the latest pip package with `pip install -U ultralytics` and see if this resolves your issue.

**Comment by Vaevin** [comment]
@Laughing-q @glenn-jocher I am a little confused. 

1) Why are the image shapes different if I pass an image list or a single image? Passing an image list resizes the images to (640, 640). However, passing a single image resizes to (480, 640).
2) I updated YOLOv8, but when I pass an image list it is still printing out multiple images of shape (1, 3, 640, 640) instead of (24, 3, 640, 640).
3) Also, @Laughing-q, where did you add a shape print in predictor.py?

![image](https://user-images.githubusercontent.com/60018242/225708029-62769fb2-d255-427d-8d43-8badb724bbd7.png)

**Comment by Laughing-q** [comment]
@Vaevin before model inference, [https://github.com/ultralytics/ultralytics/blob/701fba47700dee7a6bd8ff4d219eca91adc9d3be/ultralytics/yolo/engine/predictor.py#L172](https://github.com/ultralytics/ultralytics/blob/701fba47700dee7a6bd8ff4d219eca91adc9d3be/ultralytics/yolo/engine/predictor.py#L172)
![image](https://user-images.githubusercontent.com/61612323/225790012-77c95e09-2094-40f5-b12b-deafe62f9aae.png)

Actually you can see it's "per image at shape ..." in the log, it just shows the image shape but doesn't mean it's single inference mode.😂
![image](https://user-images.githubusercontent.com/61612323/225790751-b144a335-e708-457a-866b-f4e6663a2b3c.png)

And the difference of the shapes is from the `auto` arg in `LetterBox` which resize and pad image before inference. It'll automatically pad image to a rectangle with each side divided by 32 after resizing instead of padding to square.
[https://github.com/ultralytics/ultralytics/blob/701fba47700dee7a6bd8ff4d219eca91adc9d3be/ultralytics/yolo/data/augment.py#L448](https://github.com/ultralytics/ultralytics/blob/701fba47700dee7a6bd8ff4d219eca91adc9d3be/ultralytics/yolo/data/augment.py#L448)

**Comment by Vaevin** [comment]
@Laughing-q Mhhh strange, I did add the `print(im.shape)` in `predictor.py`, and nothing shows. But now I am realizing that the `predictory.py` script is actually not even being called when I call the `model.predict()`. Apologies for the confusion; I do not fully understand how to actually run the predictor.py when I am using this:
```
from ultralytics import YOLO
model = YOLO(model=weight_path)
```
But I am unsure. I will look into it more, but if you could provide much needed context, I'd appreciate any.

@glenn-jocher

**Comment by Vaevin** [comment]
To summarize this discussion for the next person who stumbles across this question:

- Batch inference is supported in YOLOv8 when the source is a list of images and/or batch tensor.
- Single image inference can sometimes have a similar speed to batch inference due to the use of "rect inference," which is a faster method.
- Rect inference is allowed when processing a single image, and the shapes of the images can be different.
- Batch inference involves stacking multiple images together before processing them. Rect inference is not supported if the original shapes of the images in the batch are different. In this case, the input size must be the same for all images.
- Batch inference can benefit from the speed of rect inference if the original shapes of the images in the batch are the same.
- The user's specific case encountered slower speeds with batch inference, so single image mode was recommended.

**Comment by Laughing-q** [comment]
@Vaevin nice summary! Thanks!

**Comment by zakmatt** [comment]
> @Laughing-q @glenn-jocher Thank you for replying. So, I know that in YOLOv5, when I passed in an image list just as you had provided, it does perform batch inference; I am able to confirm inference speed times increases dramatically. However, YOLOv8 does not appear to support this. Here is my code to recreate the YOLOv8 case (you'll need your own image folderpath and your own model):
> 
> # Performs batch inference by passing in a list of images as source.
> ```
> import sys
> from pathlib import Path
> file = Path(__file__).resolve()
> root = file.parents[2]  # YOLOv5 root directory
> if str(root) not in sys.path:
>     sys.path.append(str(root))  # add ROOT to PATH
> import cv2
> import numpy as np
> import os
> from ultralytics import YOLO
> 
> # This init_YOLO sets up everything we need. To change any parameters, weights, filepaths, etc, you MUST edit 
> # the  config_YOLO.yaml file.
> model = YOLO(model=weights)
> 
> weights = # Insert weights path here
> Images_folderpath = # Insert folder path path here
> 
> images = []
> for img in os.listdir(Images_folderpath):
>     images.append(cv2.imread(os.path.join(Images_folderpath, img)))
> 
> #%%
> # Create image list
> list_sz = 24
> imgs = []
> for i in range(list_sz):
>     imgs.append(images[i])
> 
> 
> #%% Pass in each image one at a time.
> inference_times = np.empty(len(imgs))
> dt_total_list = []
> for img in imgs:
>     predictions = model.predict(source=img, save=False, save_conf=True, max_det=1, show=False, stream=False, device=0)
>     
>     for prediction in predictions:
>         dt_dict = prediction.speed
>         dt = []
>         for key, value in dt_dict.items():
>             dt.append(value)
>         dt_total = sum(dt)
>         dt_total_list.append(dt_total)
>         
> # Once all 20 have been looped, add that list to our inference_times
> inference_times = np.array(dt_total_list)
> 
> print(f"Average Inference Speed: {np.mean(inference_times):.2f}")
> 
> #%% Pass in the entire imgs list
> inference_times = np.empty(len(imgs))
> dt_total_list = []
> 
> predictions = model.predict(source=imgs, save=False, save_conf=True, max_det=1, show=False, stream=False, device=0)
> 
> for prediction in predictions:
>     dt_dict = prediction.speed
>     dt = []
>     for key, value in dt_dict.items():
>         dt.append(value)
>     dt_total = sum(dt)
>     dt_total_list.append(dt_total)
>         
> inference_times = np.array(dt_total_list)
> 
> print(f"Average Inference Speed: {np.mean(inference_times):.2f}")
> ```
> 
> After running the above, I see that it still only passes 1 image at a time, even when the source is an image list. The inference speeds also confirm that they are the same between passing an image list or one image at a time.
> 
> Regarding the tensor, the tensor input also does not work as a source input.
> 
> I like the functionality of YOLOv8 and would love for it to support batch inference; I think it would be a widely regarded feature amongst the community.
`images.append(cv2.imread(os.path.join(Images_folderpath, img)))` <- here. you're adding BGR images and you should use RGBs. Swap 2 axes by adding cv2.cvtColor(img, cv2.COLOR_BGR2RGB).

**Comment by glenn-jocher** [comment]
Removing TODO as batched inference is now fully supported :)

**Comment by HaydenFaulkner** [comment]
Interestingly using torch tensors I found batch processing slower (for larger batches) than python loop iterations:
```python
from ultralytics import YOLO
import time
import torch

model = YOLO('yolov8n-seg.yaml')

for i in [2,4,8,16,32]:
    start_time = time.time()
    results = model.predict(torch.rand(i, 3, 640, 640, dtype=torch.float32), verbose=False)
    print("--- Batched (%d images) totalling %02f seconds .. average of %02f seconds per image---" % (i, time.time() - start_time, (time.time() - start_time)/i))
    start_time = time.time()
    for j in range(i):
        results = model.predict(torch.rand(1, 3, 640, 640, dtype=torch.float32), verbose=False)
    print("vs- Looped  (%d images) totalling %02f seconds .. average of %02f seconds per image---" % (i, time.time() - start_time, (time.time() - start_time)/i))
    print("---"*20)
```

produces:
```txt
--- Batched (2 images) totalling 0.701171 seconds .. average of 0.350586 seconds per image---
vs- Looped  (2 images) totalling 0.795030 seconds .. average of 0.397515 seconds per image---
------------------------------------------------------------
--- Batched (4 images) totalling 1.392624 seconds .. average of 0.348156 seconds per image---
vs- Looped  (4 images) totalling 1.449560 seconds .. average of 0.362390 seconds per image---
------------------------------------------------------------
--- Batched (8 images) totalling 3.524460 seconds .. average of 0.440558 seconds per image---
vs- Looped  (8 images) totalling 2.864456 seconds .. average of 0.358057 seconds per image---
------------------------------------------------------------
--- Batched (16 images) totalling 7.362275 seconds .. average of 0.460142 seconds per image---
vs- Looped  (16 images) totalling 5.733473 seconds .. average of 0.358342 seconds per image---
------------------------------------------------------------
--- Batched (32 images) totalling 17.338341 seconds .. average of 0.541823 seconds per image---
vs- Looped  (32 images) totalling 14.102532 seconds .. average of 0.440704 seconds per image---
------------------------------------------------------------
```
And the gap grows larger the larger the batch size. Any thoughts on why this might be the case?
Thanks

**Comment by glenn-jocher** [comment]
@HaydenFaulkner thanks for reaching out!

When using torch tensors as input, the bottleneck might lie in the conversion from tensor to numpy array before passing to the YOLOv8 model. Since numpy arrays are not natively supported by the GPU, this conversion step is likely performed on the CPU, which can be slower for larger batch sizes due to the overhead of managing larger amounts of data.

On the other hand, when using a python loop to pass in individual images, the overhead of the conversion step is reduced because you are only passing in one image at a time. As a result, the looped approach may be faster for larger batch sizes when using torch tensors as input.

Of course, these observations may depend on specific hardware configurations and software installations, and further intuitions may be gleaned by profiling the execution time of different stages of the inference pipeline.

I hope this helps! Let us know if you have any further questions or concerns.

**Comment by HaydenFaulkner** [comment]
@glenn-jocher thanks for your prompt response, this makes sense as I'm running on a Jetson Orin, I kinda expected a torch tensor to be even faster but I wrongly assumed no conversion to numpy would be made prior to getting to the model. Indeed passing in images as a list does achieve somewhat of a batching speedup (as suggested previously):
```txt
--- Batched (1 images) totalling 1.15 seconds .. average of 1.15 seconds per image---
vs- Looped  (1 images) totalling 0.91 seconds .. average of 0.91 seconds per image---
------------------------------------------------------------
--- Batched (2 images) totalling 1.21 seconds .. average of 0.60 seconds per image---
vs- Looped  (2 images) totalling 1.72 seconds .. average of 0.86 seconds per image---
------------------------------------------------------------
--- Batched (4 images) totalling 1.59 seconds .. average of 0.40 seconds per image---
vs- Looped  (4 images) totalling 3.04 seconds .. average of 0.76 seconds per image---
------------------------------------------------------------
--- Batched (8 images) totalling 2.87 seconds .. average of 0.36 seconds per image---
vs- Looped  (8 images) totalling 4.96 seconds .. average of 0.62 seconds per image---
------------------------------------------------------------
--- Batched (16 images) totalling 5.86 seconds .. average of 0.37 seconds per image---
vs- Looped  (16 images) totalling 8.19 seconds .. average of 0.51 seconds per image---
------------------------------------------------------------
--- Batched (32 images) totalling 16.45 seconds .. average of 0.51 seconds per image---
vs- Looped  (32 images) totalling 18.20 seconds .. average of 0.57 seconds per image---
```

Thanks for your insights!

**Comment by glenn-jocher** [comment]
Hi @HaydenFaulkner,

Thank you for sharing your findings and insights on the performance comparison between batched inference using torch tensors and looping through individual images. Your observations are indeed interesting and can provide valuable information for optimizing inference on different platforms.

Regarding the expectation of torch tensors being faster, it's important to note that although tensors are inherently compatible with GPU processing, there can still be overhead in the conversion step from tensors to numpy arrays. This conversion typically takes place on the CPU, and for larger batch sizes, the overhead can become more noticeable.

Additionally, on certain platforms like your Jetson Orin, passing images as a list indeed achieves some batching speedup, as each image is treated as an individual input to the model. This approach reduces the conversion overhead and can lead to faster inference times, especially for larger batch sizes.

Thank you once again for sharing your results and insights. Your findings can be beneficial for others working on similar platforms, and it's great to see the community exploring and optimizing the performance of YOLOv8. If you have any further questions or observations, feel free to reach out. 

Happy coding!
